cmd/agent/app/throttling/throttler.go

```diff
@@ -0,0 +1,294 @@
+	AccountOptionOverrides map[string]*AccountOptions
+
+	// ClientMaxBalance defines the maximum balance a client may maintain before
+	// further Withdraw calls return zero.
+	ClientMaxBalance float64
+
+	// TTL defines the time to await further updates before purging entry for
+	// all accounts and clients.
+	TTL time.Duration
+	// PurgeInterval defines the amount of time to wait between purge calls.
+	PurgeInterval time.Duration
+}
+
+// Throttler manages the relationship between service instance clients and
+// backend storage credits. Each client must withdraw credits in order to submit
+// trace data to the Jaeger agent. These credits are generated on behalf of the
+// service, of which there can be many instances, and thus many clients.
+type Throttler struct {
+	sync.Mutex
+	accounts  map[string]*account
+	clients   map[string]*client
+	options   ThrottlerOptions
+	timeNow   func() time.Time // For testing
+	ticker    *time.Ticker
+	done      chan bool
+	waitGroup sync.WaitGroup
+}
+
+// NewThrottler creates a new throttler and returns it to the caller.
+// options should specify the specific throttler options to use.
+func NewThrottler(options ThrottlerOptions) *Throttler {
+	t := &Throttler{
+		accounts: map[string]*account{},
+		clients:  map[string]*client{},
+		options:  options,
+		timeNow:  time.Now,
+	}
+	t.ticker = time.NewTicker(t.options.PurgeInterval)
+	t.done = make(chan bool)
+	t.waitGroup.Add(1)
+	go func() {
+		defer t.waitGroup.Done()
+		for {
+			select {
+			case <-t.ticker.C:
+				t.purgeExpired()
+			case <-t.done:
+				return
+			}
+		}
+	}()
+	return t
+}
+
+// Withdraw deducts as many credits from the service on behalf of the client
+// without exceeding MaxClientBalance. The client is identified by a unique
+// clientID string. The credits are deducted from the specific operation on
+// the service if present, otherwise from the default credit pool.
+func (t *Throttler) Withdraw(serviceName string, clientID string, operationName string) float64 {
+	t.Lock()
+	defer t.Unlock()
+	c := t.findOrCreateClient(clientID)
+	a := t.findOrCreateAccount(serviceName)
+	balance := c.perOperationBalance[operationName]
+	maxWithdrawal := t.options.ClientMaxBalance - balance
+	credits := a.Withdraw(operationName, maxWithdrawal)
+	now := t.timeNow()
+	c.updateTime = now
+	a.updateTime = now
+	c.perOperationBalance[operationName] += credits
+	return credits
+}
+
+// Spend spends credits already allocated to this client on a given operation.
+func (t *Throttler) Spend(serviceName string, clientID string, operationName string, credits float64) error {
+	t.Lock()
+	defer t.Unlock()
+	c := t.findOrCreateClient(clientID)
+	c.updateTime = t.timeNow()
+	err := c.Spend(operationName, credits)
+	if err != nil {
+		err = errors.New(err.Error() + ", serviceName " + serviceName)
+	}
+	return err
+}
+
+// Close closes the throttler and stops all background goroutines.
+func (t *Throttler) Close() error {
+	t.done <- true
+	close(t.done)
```

Comment:
afaik, `close(t.done)` will cause the for loop to exit, the `t.done <- true` isn't required.

Reply:
Good point.